pkg/scalers/activemq_scaler.go

```diff
@@ -0,0 +1,268 @@
+
+	if len(v["brokerName"][0]) == 0 {
+		return meta, fmt.Errorf("no brokerName given: %s", meta.restAPITemplate)
+	}
+	meta.brokerName = v["brokerName"][0]
+
+	return meta, nil
+}
+
+func (s *activeMQScaler) getMonitoringEndpoint() string {
+	replacer := strings.NewReplacer("<<managementEndpoint>>", s.metadata.managementEndpoint,
+		"<<brokerName>>", s.metadata.brokerName,
+		"<<destinationName>>", s.metadata.destinationName)
+
+	monitoringEndpoint := replacer.Replace(s.metadata.restAPITemplate)
+
+	return monitoringEndpoint
+}
+
+func (s *activeMQScaler) getQueueMessageCount(ctx context.Context) (int, error) {
+	var monitoringInfo *activeMQMonitoring
+	var queueMessageCount int
+
+	client := s.httpClient
+	url := s.getMonitoringEndpoint()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+
+	// Add HTTP Auth and Headers
+	req.SetBasicAuth(s.metadata.username, s.metadata.password)
+	req.Header.Set("Content-Type", "application/json")
+
+	if err != nil {
+		return -1, err
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return -1, err
+	}
+
+	defer resp.Body.Close()
+
+	if err := json.NewDecoder(resp.Body).Decode(&monitoringInfo); err != nil {
+		return -1, err
+	}
+	if resp.StatusCode == 200 && monitoringInfo.Status == 200 {
+		queueMessageCount = monitoringInfo.MsgCount
+	} else {
+		return -1, fmt.Errorf("ActiveMQ management endpoint response error code : %d %d", resp.StatusCode, monitoringInfo.Status)
+	}
+
+	activeMQLog.V(1).Info(fmt.Sprintf("ActiveMQ scaler: Providing metrics based on current queue size %d queue size limit %d", queueMessageCount, s.metadata.targetQueueSize))
+
+	return queueMessageCount, nil
+}
+
+// GetMetricSpecForScaling returns the MetricSpec for the Horizontal Pod Autoscaler
+func (s *activeMQScaler) GetMetricSpecForScaling(context.Context) []v2beta2.MetricSpec {
+	targetMetricValue := resource.NewQuantity(int64(s.metadata.targetQueueSize), resource.DecimalSI)
+	externalMetric := &v2beta2.ExternalMetricSource{
+		Metric: v2beta2.MetricIdentifier{
+			Name: GenerateMetricNameWithIndex(s.metadata.scalerIndex, s.metadata.metricName),
+		},
+		Target: v2beta2.MetricTarget{
+			Type:         v2beta2.AverageValueMetricType,
+			AverageValue: targetMetricValue,
+		},
+	}
+	metricSpec := v2beta2.MetricSpec{
+		External: externalMetric, Type: externalMetricType,
+	}
+	return []v2beta2.MetricSpec{metricSpec}
+}
+
+func (s *activeMQScaler) GetMetrics(ctx context.Context, metricName string, metricSelector labels.Selector) ([]external_metrics.ExternalMetricValue, error) {
+	queueSize, err := s.getQueueMessageCount(ctx)
+	if err != nil {
+		return []external_metrics.ExternalMetricValue{}, fmt.Errorf("error inspecting ActiveMQ queue size: %s", err)
+	}
+
+	metric := external_metrics.ExternalMetricValue{
+		MetricName: metricName,
+		Value:      *resource.NewQuantity(int64(queueSize), resource.DecimalSI),
+		Timestamp:  metav1.Now(),
+	}
+
+	return append([]external_metrics.ExternalMetricValue{}, metric), nil
```

Comment:
```suggestion
	return []external_metrics.ExternalMetricValue{metric}, nil
```

Reply:
I didn't understand this and previous suggestion. I checked other scalers to understand but they've also written this way. Could you please explain?

Reply:
certainly!

regarding the previous suggestion, you can pass or return `nil` where a slice is expected, and the compiler will do the "right thing" (treat it as an empty slice)

Regarding this one, the `append` approach works, but I think it's an unnecessary step since it creates an empty slice and immediately adds one element to the end of it. The code I suggested creates the slice with that element inside it. I don't recall how the compiler treats the `append` code, but I do believe this change makes the code easier to read

Reply:
Thank you for detailed explanation.